### DISCUSSION TITLE
Manually highlighting a region of a buffer using an explicit treesitter parser

### DISCUSSION THREAD

**Discussion by chipsenkbeil** [neovim/neovim#27940] [Q&A] [answered]
I have a scenario where I am creating a custom buffer filetype and generating the content within the buffer. For parts of the buffer, I want to highlight using treesitter. Is there a way for me to load up a treesitter parser for a specific language, parse a region of a buffer, and apply highlights?

The concrete example is I'm writing a port of [org-roam](https://www.orgroam.com/) which is getting close to complete, but trying to represent the org-roam buffer has proven challenging. I want to provide a preview of each backlink and provide orgmode syntax highlighting to the sections that are previews.

Looking around, I can't find an example or discussion about this.

```
Title

Backlinks (4):
▶ First Document @ Line 5
▼ Second Document @ Line 5
- Preview contents
- Of the document
- where a link would be
▶ Third Document @ Line 8
▼ Third Document @ Line 45
This is a paragraph that contains
the link in question
```

**Comment by justinmk** [comment, marked as answer]
In general to highlight a range of text the main function ins `nvim_buf_set_extmark()`. But I'm not sure how to combine that with a treesitter parser.

`:help vim.treesitter.get_string_parser()` seems related, but I think it's not bound to a buffer.

For custom highlighting start with `:help :EditQuery` https://neovim.io/doc/user/treesitter.html#vim.treesitter.query.edit() , you could define a query that enables highlighting only for a certain section of a document. However, doing that dynamically might be weird.

**Reply by chipsenkbeil** [reply]
Out of curiosity, why is `nvim_buf_set_extmark` preferred over `nvim_buf_add_highlight`? The latter is what I use today to perform all of the highlights throughout the plugin.

I can certainly give the query route a go, but not optimistic on my success there. 

Ideally, I would like to have the option of specifying one or more ranges within a buffer to highlight when using `vim.treesitter.start`. I guess the closest is trying to devise a query that can separate out orgmode from non-orgmode content.

**Reply by chipsenkbeil** [reply]
Or as an alternative to `vim.treesitter.start`, have something like `vim.treesitter.highlights_for("some text")` where I'd get a list of highlight groups and ranges representing the line, start column, and end column for each group. Because I could take those and then apply them with an offset (e.g. 3 lines down) to the buffer that contains the text.
